linera-service/src/linera-exporter/runloops/block_processor.rs

```diff
@@ -0,0 +1,461 @@
+                    hash,
+                )
+            }),
+            block: block_id,
+            siblings: block
+                .body
+                .incoming_bundles
+                .iter()
+                .map(BlockId::from_incoming_bundle)
+                .collect::<Vec<_>>(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::time::Duration;
+
+    use linera_base::{
+        crypto::CryptoHash,
+        data_types::{Round, Timestamp},
+        identifiers::ChainId,
+    };
+    use linera_chain::{
+        data_types::{BlockExecutionOutcome, IncomingBundle, MessageBundle},
+        test::{make_child_block, make_first_block, BlockTestExt},
+        types::{CertificateValue, ConfirmedBlock, ConfirmedBlockCertificate},
+    };
+    use linera_client::config::LimitsConfig;
+    use linera_sdk::test::MessageAction;
+    use linera_storage::{DbStorage, Storage};
+    use linera_views::memory::MemoryStore;
+    use tokio::sync::mpsc::unbounded_channel;
+    use tokio_util::sync::CancellationToken;
+
+    use crate::{
+        common::BlockId, runloops::BlockProcessor, storage::BlockProcessorStorage,
+        ExporterCancellationSignal,
+    };
+
+    #[tokio::test]
+    async fn test_topological_sort() -> anyhow::Result<()> {
+        let (tx, rx) = unbounded_channel();
+        let storage = DbStorage::<MemoryStore, _>::make_test_storage(None).await;
+        let (block_processor_storage, exporter_storage) =
+            BlockProcessorStorage::load(storage.clone(), 0, 0, LimitsConfig::default()).await?;
+        let mut block_processor = BlockProcessor::new(block_processor_storage, tx.clone(), rx);
+        let token = CancellationToken::new();
+        let signal = ExporterCancellationSignal::new(token.clone());
+        let (block_ids, state) = make_state(&storage).await;
+        for id in block_ids {
+            let _ = tx.send(id);
+        }
+
+        tokio::select! {
+            _ = tokio::time::sleep(Duration::from_secs(10)) => {},
+            _ = block_processor.run_with_shutdown(signal, 5) => {},
+        }
+
+        // oredered pair of (chain_id, block_height)
+        let expected_state = [
+            (1, 0),
+            (0, 0),
+            (0, 1),
+            (1, 1),
+            (1, 2),
+            (0, 2),
+            (0, 3),
+            (1, 3),
+        ];
+
+        for (i, (x, y)) in expected_state.into_iter().enumerate() {
+            let hash = exporter_storage.get_block(i).await?.hash();
+            assert_eq!(hash, state[x][y]);
+        }
+
+        Ok(())
+    }
+
+    async fn make_state<S: Storage>(storage: &S) -> (Vec<BlockId>, Vec<Vec<CryptoHash>>) {
+        let mut notifications = Vec::new();
+
+        let chain_id_a = ChainId(CryptoHash::test_hash("0"));
+        let chain_id_b = ChainId(CryptoHash::test_hash("1"));
+
+        let mut chain_a = Vec::new();
+        let mut chain_b = Vec::new();
+
+        for i in 0..4 {
+            if 0 == i {
```

Comment:
```suggestion
            if i == 0 {
```
That reads weird otherwise.

Reply:
Still not addressed.